feat(build): add "vite:emit" output plugin

aleclarson opened the pull request:
- Run post-build hooks before the `writeBundle` phase
- Allow post-build hooks to write assets by mutating the `build.assets` array
- Empty the `outDir` after the build succeeds (right before the `writeBundle` phase)
- Write the index.html in the `writeBundle` phase
- Copy the public dir in the `writeBundle` phase
- Respect `emitAssets: true` even if `write` is false
- Respect `emitIndex: true` even if `write` is false

---

- Fixes `vite-plugin-legacy`, which was broken by https://github.com/vitejs/vite/pull/957#issuecomment-717926853
- Reverts https://github.com/vitejs/vite/commit/5e7c309baa16b78fbfdfbf3fffff7af775c26047, which I tacked onto another PR without understanding the `emitIndex` option's purpose, which is to prevent `index.html` rendering entirely, not just to prevent writing. (emit !== write)

aleclarson committed:
feat(build): add "vite:emit" output plugin

- Run post-build hooks before the `writeBundle` phase
- Allow post-build hooks to write assets by mutating the `build.assets` array
- Empty the `outDir` after the build succeeds (right before the `writeBundle` phase)
- Write the index.html in the `writeBundle` phase
- Copy the public dir in the `writeBundle` phase
- Respect `emitAssets: true` even if `write` is false
- Respect `emitIndex: true` even if `write` is false

aleclarson commented:
Looks like the `vite:emit` plugin is running before others, eg: the CSS plugin. Fixing it...

aleclarson committed:
fix: ensure "vite:emit" plugin runs last

and replace `Build.bundle` with `Build.options`

aleclarson committed:
fix(config): disallow `sourcemap: hidden` value

No point in generating an external sourcemap without a corresponding sourceMappingURL.

And this commit allows `configureBuild` hooks to pass the `sourcemap` value to other tools (like Babel) that do not recognize "hidden" as a valid value.

aleclarson commented:
Okay, I worked around the issue in my previous comment.

The `Build.bundle` property is replaced by the `Build.options` property, which `configureBuild` hooks can use to customize the input/output options of each build. Then we append the `vite:emit` plugin so it runs last.

LGTM 👍

aleclarson committed:
fix(types): make `Build.id` readonly

We don't want `configureBuild` hooks to rename the main build, because that's how we detect the main build in the "vite:emit" plugin.

aleclarson committed:
refactor: arguments of `createEmitPlugin`

aleclarson committed:
refactor: `indexPath` is always a string

aleclarson committed:
fix: ignore "vite:emit" plugin copied from other build

aleclarson committed:
fix: only clear `outDir` and copy `publicDir` once

aleclarson commented:
Whacked the last few bugs

yyx990803 reviewed (commented):
Looks good, but I think we should add a test case for multi builds.

aleclarson committed:
test: multi build

aleclarson commented:
@yyx990803 I stubbed out a basic test for multi builds: https://github.com/vitejs/vite/pull/977/commits/55187131a0793b481f7ecd29306211e523076d26

No tests yet for:
- mutating the `config` argument in `configureBuild` hook
- pushing to `build.assets` in a post-build hook

aleclarson committed:
feat: allow builds that don't render html

- The `Build` interface was reworked:
  - It extends the `InputOptions` interface from Rollup, since all properties
    other than `options` were removed.
  - The `output` property is now required.
  - The `onResult` property was added, which runs before global post-build hooks
    returned by `configureBuild` functions.

- Post-build hooks now receive a `BuildResult` object, instead of a `Build` object.
  Note: The `Build` object can still be accessed with the `build` property.

- The `renderIndex` function is only called when a build contains the main "index.html"
  path (known as `entry` in Vite config) in its `build.input` value.

- The output path of the `renderIndex` result is inferred from the `build.input` value,
  which is normalized into an object whose keys are output paths.

- Written files are logged between builds, instead of waiting for all builds to complete.

- Simplify the `createEmitPlugin` function. It now takes only an `emit` function, which
  allows closing over local variables instead of passing them in.

- The `onwarn` option is now inherited by all builds, but can be overridden.

aleclarson commented:
I pushed a new commit with a bunch of changes: https://github.com/vitejs/vite/pull/977/commits/c894e2dcb2db61d895e235f24906f789d1775d85

The changes are listed in the commit message. 
Testing them out now.

Commit:
chore: update lockfile

aleclarson committed:
refactor: remove `normalizeInput` function

To keep things simple, only call the `renderIndex` function when the `build.input` value
equals "index.html". And let `build.output.file` override the default output path of
"{outDir}/index.html".

aleclarson commented:
Pushed https://github.com/vitejs/vite/pull/977/commits/0032d93c3431cd194dc0bd0d4c66f70f7fb93fcf, which simplifies how the `index.html` output path is resolved by replacing the `normalizeInput` function I added earlier in this PR. More details in the commit itself.

aleclarson committed:
fix: revert `config.entry` usage

yyx990803 committed:
Merge branch 'master' into feat/vite-emit

yyx990803 committed:
refactor: move assets removal to emit plugin

aleclarson committed:
fix: remove unused argument

pd4d10 commented:
This PR seems break `config.entry`...
